computer-vision/robot_tracking.cpp

```diff
@@ -0,0 +1,167 @@
+        cv::read(fs["camera_matrix"], camera_matrix);
+        cv::read(fs["distortion_coefficients"], distortion_coefficients);
+    }
+};
+
+enum WeathiervaneDirection {
+    NORTH = 0,
+    SOUTH = 1
+};
+
+WeathiervaneDirection weathervane_direction(cv::Vec3d rvec, cv::Vec3d tvec)
+{
+    cv::Mat R;
+    cv::Rodrigues(rvec, R);
+    cv::Vec3d e2 = {0, 1, 0};
+    cv::Vec3d v = cv::Mat(R * e2);
+
+    // check if facing down
+    if (v[1] > 0) {
+        return SOUTH;
+    } else {
+        return NORTH;
+    }
+}
+
+ComputerVisionResult processImage(cv::Mat& image, RobotTrackingSettings& settings, bool display = false)
+{
+    float markerLength = 0.1;
+    const int dictionaryId = 1; // DICT_4X4_100=1
+
+    const cv::Mat& M = settings.camera_matrix;
+    const cv::Mat& D = settings.distortion_coefficients;
+
+    cv::Ptr<cv::aruco::DetectorParameters> detectorParams = cv::aruco::DetectorParameters::create();
+    cv::Ptr<cv::aruco::Dictionary> dictionary =
+        cv::aruco::getPredefinedDictionary(cv::aruco::PREDEFINED_DICTIONARY_NAME(dictionaryId));
+
+    std::vector<int> ids;
+    std::vector<std::vector<cv::Point2f>> corners, rejected;
+    std::vector<cv::Vec3d> rvecs, tvecs;
+
+    // detect markers and estimate pose
+    cv::aruco::detectMarkers(image, dictionary, corners, ids, detectorParams, rejected);
+    if (ids.size() > 0) {
+        cv::aruco::estimatePoseSingleMarkers(corners, markerLength, M, D, rvecs,
+                                             tvecs);
+    }
+
+    for (unsigned int i = 0; i < ids.size(); i++) {
+        std::cout << ids[i] << std::endl;
+        std::cout << rvecs[i] << std::endl;
+        std::cout << tvecs[i] << std::endl;
+    }
+
+    int dir = -1;
+    for (size_t i = 0; i < ids.size(); i++) {
+        if (ids[i] == ARUCO_WEATHERVANE) {
+            dir = weathervane_direction(rvecs[i], tvecs[i]);
+            break;
+        }
+    }
+
+    switch (dir) {
+        case NORTH:
+            printf("NORTH\n");
+            break;
+        case SOUTH:
+            printf("SOUTH\n");
+            break;
+    }
+
+    if (display) {
+        cv::Mat imageCopy;
+        // draw results
+        image.copyTo(imageCopy);
+
+        std::cout << "found " << ids.size() << " markers." << std::endl;
+        if (ids.size() > 0) {
+            cv::aruco::drawDetectedMarkers(imageCopy, corners, ids);
+        }
+
+        for (unsigned int i = 0; i < ids.size(); i++) {
+            cv::aruco::drawAxis(imageCopy, M, D, rvecs[i], tvecs[i],
+                                markerLength * 0.5f);
+        }
+
+        // cv::imwrite("out.jpg", imageCopy);
+        cv::imshow("out", imageCopy);
+    }
+
+    ComputerVisionResult result;
+    result.set_weather_vane_orientation(ComputerVisionResult::NORTH);
+    return result;
+}
+
+int main(int argc, char* argv[])
+{
+    logging_init();
+    if (argc < 2) {
+        ERROR("usage: %s config.yaml image.jpg\n", argv[0]);
+        return -1;
```

Comment:
```suggestion
        return EXIT_FAILURE;
```
or
```suggestion
        throw("usage: %s config.yaml <image.jpg>\n", argv[0]);
```

Reply:
Yes, need to cleanup the code. This is more of a proof of concept

Reply:
This was built upon an opencv example. Focus was on functionality not style. We'll refactor later or in another PR.

Reply:
I'd still refactor here, it's OK, I just don't care it's copied from an example, I'd like it cleaned up before merge